### ISSUE TITLE
OpenSSL::X509::Store support broken in JRuby

### ISSUE THREAD

**Issue by quixoten** [nahi/httpclient#276]
It looks like the latest version of httpclient broke the use of a custom cert_store when running under JRuby. It doesn't look like there's a straightforward way to fix this with the separated SSLSocket implementation.

**Comment by alex** [comment]
How is this manifesting for you? I'm seeing a `sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target` error on 2.7 for something that works fine on 2.6

**Comment by quixoten** [comment]
I see the same thing.

**Comment by nahi** [comment]
@quixoten You mean that you have `OpenSSL::X509::Store` and you want to let HTTPClient use the store for certificate verification, right? Sorry, I forgot to consider this problem when I switch SSL implementation for JRuby... HTTPClient needs certificates but OpenSSL::X509::Store does not provide any method for it.

Please let me know what library you're using. I want to try to find a workaround for this problem. Anyway sorry for the inconvenience this incompatible change causes.

**Comment by alex** [comment]
@nahi https://github.com/department-of-veterans-affairs/connect_vbms/blob/master/src/vbms/client.rb#L49-L60 is the code that doesn't work (specifically the `if @key` path)

**Comment by nahi** [comment]
@alex Thanks! But the code is different what I was worrying about (SSLConfig#set_cert_store.) The code should work but I may be misunderstanding something. Can you provide the file for CONNECT_VBMS_CACERT to me?

**Comment by alex** [comment]
Unfortunately not, though the code is open source, the API it communicates
with and everything reltaed is a properietary internal system.

`CONNECT_VBMS_CACERT` really is just a normal PEM x509 cert though, if you
have questions about it I can probably answer them.

On Sun, Dec 20, 2015 at 11:22 AM, Hiroshi Nakamura <notifications@github.com
> wrote:

> @alex <https://github.com/alex> Thanks! But the code is different what I
> was worrying about (SSLConfig#set_cert_store.) The code should work but I
> may be misunderstanding something. Can you provide the file for
> CONNECT_VBMS_CACERT to me?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nahi/httpclient/issues/276#issuecomment-166133373>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by nahi** [comment]
@alex OK, I'll try to reproduce by other certificates.

I tried with https://cacert.org. With "Root Certificate (PEM Format)" here at http://www.cacert.org/index.php?id=3 I can connect to https://cacert.org.
```
% jruby -J-Djsse.enableSNIExtension=false -Ilib -rhttpclient -e 'p HTTPClient.new { |c| c.ssl_config.set_trust_ca("/Users/nahi/Downloads/root.crt") }.get("https://cacert.org").headers'
{"Date"=>"Sun, 20 Dec 2015 16:26:21 GMT", "Server"=>"Apache/2.2.22 (Debian)", "Strict-Transport-Security"=>"max-age=31536000", "Set-Cookie"=>"cacert=khda6o3gamtqrmfv8bfhafkv42; path=/; secure", "Expires"=>"Thu, 19 Nov 1981 08:52:00 GMT", "Cache-Control"=>"no-store, no-cache, must-revalidate, post-check=0, pre-check=0", "Pragma"=>"no-cache", "location"=>"https://www.cacert.org", "Vary"=>"Accept-Encoding", "Content-Length"=>"0", "Content-Type"=>"text/html; charset=iso-8859-1"}
```

So set_trust_ca should work.

Could you provide full stack trace and error message? How many certificates in CONNECT_VBMS_CACERT file? How error message changes when you remove these client authentication lines? https://github.com/department-of-veterans-affairs/connect_vbms/blob/master/src/vbms/client.rb#L51-L53

Thanks!

**Comment by nahi** [comment]
Anyone who found JRuby + SSL custom verification trouble please let me know here (or privately at nahi@ruby-lang.org.) I want to fix any problem I could introduce from 2.7.x.

**Comment by petr666** [comment]
I have the same problem. I'm attaching simple test script:
[x509_store_problem.txt](https://github.com/nahi/httpclient/files/363385/x509_store_problem.txt)

**Comment by nahi** [comment]
@petr666 Thank you very much. Let me check...

**Comment by nahi** [comment]
Confirmed the cause. I'm trying to fix this at https://github.com/nahi/httpclient/pull/317

**Comment by petr666** [comment]
I've tried patch from #317 and it seems to be working, looking forward for next release :o) However I've noticed one more problem. In 2.8 with JRuby it is not possible to disable SSL validation using "client.ssl_config.verify_mode = OpenSSL::SSL::VERIFY_NONE". When set (e.g. in my test case instead of adding custom certificate to store) it fails with exactly same error (including backtrace). It seems disabling validation is ignored in JRuby code (again it works in Ruby, and also JRuby with httpclient 2.6).
